File: examples/19-ddd-hexagonal/dotnet/OrderService/Domain/PlaceOrderUseCase.cs
namespace OrderService.Domain;

public class PlaceOrderUseCase
{
    private readonly IOrderRepository _repo;
    private readonly IEventPublisher _events;

    public PlaceOrderUseCase(IOrderRepository repo, IEventPublisher events)
    {
        _repo = repo;
        _events = events;
    }

    public async Task<Order> ExecuteAsync(PlaceOrderCmd cmd)
    {
        var order = Order.Create(cmd);
        await _repo.SaveAsync(order);
        await _events.PublishAsync(new OrderPlaced(order.Id, order.Sku, order.Quantity));
        return order;
    }
}

File: examples/19-ddd-hexagonal/quarkus/order-service/src/main/java/com/cndp/order/domain/PlaceOrderUseCase.java
package com.cndp.order.domain;

public class PlaceOrderUseCase {

    private final OrderRepository repo;
    private final EventPublisher events;

    public PlaceOrderUseCase(OrderRepository repo, EventPublisher events) {
        this.repo = repo;
        this.events = events;
    }

    public Order execute(PlaceOrderCmd cmd) {
        Order order = Order.create(cmd);
        repo.save(order);
        events.publish(new OrderPlaced(order.id(), order.sku(), order.quantity()));
        return order;
    }
}
